### PR TITLE
Support for addresses with staking keys #103 [obsolete].

### PR DESCRIPTION
**Pull request by zmrocze**
- Add stake keys to BpiWallet.
- Generate stake keys in `addWallets`
- `withStakeKeys` to add stake keys to `TestWallets`
- Refactor key paths.
- Set `pcOwnStakePubKeyHash` for BPI.
- Pass `WalletInfo`'s to contracts instead of `PubKeyHash`'s.
- Update tests with `onEnterpriseAddresses` helper.
-  TODO: Add tests.
- TODO: Add another interface with lookups. 
- TODO: Add superduper interface with polyvariadic `withContract`.

### CODE REVIEW

**Comment by zmrocze** [comment]
I haven't added any tests yet. All the tests are left unchanged, that is they are initialized with enterprise addresses (no stake). Some of them rely on running with enterprise address.

**Comment by zmrocze** [comment]
TODO: Update changelog, but once we're settled with interface change

File: src/Test/Plutip/Internal/BotPlutusInterface/Run.hs [hunk 1]
```diff
@@ -77,15 +78,15 @@ runContract_ ::
...
+  WalletInfo ->
   Contract w s e a ->
   m ()
 runContract_ e w c = void $ runContract e w c
 
 runContract ::
   (ToJSON w, Monoid w, MonadIO m) =>
   ClusterEnv ->
-  BpiWallet ->
+  WalletInfo ->
```
Comment (mikekeke) [inline src/Test/Plutip/Internal/BotPlutusInterface/Run.hs:89]: Changing type of this argument will break some docs.
Like [here](https://github.com/mlabs-haskell/plutip/blob/master/local-cluster/README.md?plain=1#L51-L57) and [here](https://github.com/mlabs-haskell/plutip/blob/master/docs/interactive-plutip.md?plain=1#L151-L154).

Not sure now how to arrange it better. Maybe `addSomeWallet` can return `WalletInfo`, so we sort of take out relation between "wallets that cluster uses" and BPI, as BPI is an implementation detail.

File: test/Spec/Integration.hs [hunk 2]
```diff
@@ -103,13 +104,13 @@ test =
...
           (initAda [100])
-          (withContract $ \[pkh1] -> payTo pkh1 (-10_000_000))
+          (withContract . onEnterpriseWallets $ \[pkh1] -> payTo pkh1 (-10_000_000))
           [shouldFail]
       , -- Tests with wallet's Value assertions
         assertExecution
           "Pay from wallet to wallet"
           (initAda [100] <> initAndAssertAda [100, 13] 123)
-          (withContract $ \[pkh1] -> payTo pkh1 10_000_000)
+          (withContract . onEnterpriseWallets $ \[pkh1] -> payTo pkh1 10_000_000)
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply test/Spec/Integration.hs:113]: i was thinking of it only as a helper to quickly adjust existing tests with little effort and not something that should be used in newly written tests. I would leave that in the exports, though as our tests double as examples for documentation, i see that i should actually go through them and update to use new api.

File: src/Test/Plutip/Contract/Types.hs [hunk 3]
```diff
@@ -3,10 +3,8 @@
 module Test.Plutip.Contract.Types (
   TestContractConstraints,
   TestContract (..),
-  TestWallets (TestWallets, unTestWallets),
-  TestWallet (..),
-  compareValuesWith,
-  ValueOrdering (..),
+  WalletInfo (..),
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply src/Test/Plutip/Contract/Types.hs:6]: right, I would hide the constructor

**Comment by zmrocze** [comment]
I reverted this pr to the last green commit and pr'ed latest changes instead to #136.

This is because my intention turned out to be impossible to implement ;/ more in that pr. We need to go with aproaches like the one at current revision.

**Comment by mikekeke** [comment]
Subsumed by #136

**Pull request closed without merging**